### PR TITLE
Enable the ProgressBar with the safe wallet

### LINKED ISSUE #794: Enable Orders execution progress bar for SC wallets
Related to #609

For now, progress bar is disabled for SC wallets due to a lot of issues (discussed in https://altoros.slack.com/archives/C03902HBL06/p1656947438786999 https://cowservices.slack.com/archives/C03902HBL06/p1656947438786999 )

It would be great to find a solution how to fix all related issues and enable it.

### PR DESCRIPTION
**Pull request by henrypalacios**
# Summary
Closes #794

Since the safe wallet API has a **delay** to report the **execution date**, it is proposed to update the `executionDate` when listening to the [ExecutionSuccess](https://github.com/safe-global/safe-contracts/blob/main/contracts/GnosisSafe.sol#L49) event of the Safe Contract.

⚠️ WIP ⚠️ 
- [x] Create a watcher to listen `ExecutionSuccess`
- [x] Create an updater that runs every time an order is created with a wallet safe
- [x] Overwrite `executionDate` in signed Order
- [x] Verify that the **ProgressBar** is started when the `executionDate` is updated
- [x] Avoid **ProgressBar** jumping when the state is delayed
- ~~Find out why in expert mode order falls as [RequestProviderError](https://github.com/cowprotocol/cowswap/blob/794-enable-progressbar/src/custom/utils/misc.ts#L145) when an order is placed with the same parameters~~

## To Test
- With a multi-signature wallet from the safe open to place an order
- Make the required signatures. 

The modal with the **progressBar** should start after listening to the execution event launched by the safe contract.

### CODE REVIEW

**Comment by henrypalacios** [comment]
Some notes:

1. I noticed that a query is triggered to update the transaction status with multi-signature on average every 2seconds.
SCR-20220928-gy1

2. When reviewing some mined transaction I have seen when querying the [safe-transactions API](https://safe-transaction.goerli.gnosis.io/api/v1/multisig-transactions/0x6f833fe3da69c6d7b17f9746eec0a4ed3881222cb4780a79df0242d27f962727/) how the `executionDate` field has been `null` for at least 40min. **😲 Big delay**

3. In addition, while @ramirotw was testing with me he created a multi-signature wallet and when he consulted them he came across NO FOUND.



🤔  We suspect it is something in the **Goerli** network. Maybe you know something about it @alfetopito

File: src/custom/abis/SafeWalletContract.json [hunk 1]
```diff
@@ -0,0 +1,1138 @@
+[
+    {
+      "anonymous": false,
```
Comment (alfetopito) [inline src/custom/abis/SafeWalletContract.json:3]: Please pick only the methods that we actually need to reduce the JSON size

**Comment by alfetopito** [comment]
I've asked Uxio from the Safe team to take a look at the delay. 40 min sounds slightly bad 😅

**Comment by alfetopito** [comment]
According to Uxio they had some node issues which could be related. Let's keep an eye if that ever happens again

File: src/custom/state/gnosisSafe/updater.tsx [hunk 2]
```diff
@@ -1,9 +1,16 @@
...
 import { useIsGnosisSafeWallet } from 'hooks/useWalletInfo'
 import { getSafeInfo } from 'api/gnosisSafe'
 import { useWeb3React } from '@web3-react/core'
 import { gnosisSafeAtom } from 'state/gnosisSafe/atoms'
 import { useUpdateAtom } from 'jotai/utils'
+import { useSafeWalletContract } from 'hooks/useContract'
+import { usePendingOrders, useUpdatePresignGnosisSafeTx } from '../orders/hooks'
+import { Order } from 'state/orders/actions'
+import usePrevious from '@src/hooks/usePrevious'
+import { whenWasMined } from '@src/custom/utils/blocks'
```
Comment (alfetopito) [inline src/custom/state/gnosisSafe/updater.tsx:13]: @src imports

```suggestion
import usePrevious from 'hooks/usePrevious'
import { whenWasMined } from 'utils/blocks'
```

**Review by alfetopito — COMMENTED** [review]
Couple of code comments, did not check the behaviour yet

File: src/custom/state/orders/reducer.ts [hunk 3]
```diff
@@ -187,7 +188,10 @@ export default createReducer(initialState, (builder) =>
 
       const orderObject = getOrderById(state, chainId, orderId)
       if (orderObject) {
-        orderObject.order.presignGnosisSafeTx = safeTransaction
+        orderObject.order.presignGnosisSafeTx = noOverwriteExecutionDateIfExists(
+          orderObject.order.presignGnosisSafeTx,
+          safeTransaction
+        )
```
[the remark this answers is not in the document]
Comment (henrypalacios) [reply src/custom/state/orders/reducer.ts:194]: Yes, it sounds good tome. I'll go that way.

Ramiro told me that maybe I should not use this reducer but a separate state with just `executionDate` managed with `Jotai`. WDYT about it @alfetopito?

**Comment by elena-zh** [comment]
Hey @henrypalacios , great job!
Progress bar works nicely besides these issues:
1. I see how it is [running ](https://watch.screencastify.com/v/3BNNIbET9URNT23s6n9Y)back to the start when an order is signed and is executed (both on confirmation and account modals)
2. When an order is in 'Signing' state, I see the message 'Order Submitted' on the confirmation modal. Maybe we could change it with 'Order Signing' state and show a 'frozen' (not moving) progress bar until the order is signed?
Anyways, not sure about the 'frozen' progress bar. @alfetopito , @henrypalacios , @ramirotw , WDYT?
image

3. I can't place another order with the same order details once the previous order is traded: 
place after the order
**To reproduce:** 
1. Place an order when connected to safe and wait till it is executed and traded
2. DO NOT change values in From/To fields and press Swap button once again.

File: src/custom/state/orders/reducer.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/custom/state/orders/reducer.ts:194]: Yes, I remember now that he mentioned it.
If you think it'll make the code cleaner go for it.

**Comment by henrypalacios** [comment]
> 1. I see how it is [running ](https://watch.screencastify.com/v/3BNNIbET9URNT23s6n9Y)back to the start when an order is signed and is executed (both on confirmation and account modals)

I am aware of this, I am working on it.

> 2. When an order is in 'Signing' state, I see the message 'Order Submitted' on the confirmation modal. Maybe we could change it with 'Order Signing' state and show a 'frozen' (not moving) progress bar until the order is signed?

When I saw it I thought of showing the signing icon 
image

> 3. I can't place another order with the same order details once the previous order is traded:

I will look at it and get back with comments!

**Comment by henrypalacios** [comment]
@elena-zh 

1. 

I have solved that case.

2. 
Now I've added a simple message, but maybe the most convenient is to be consistent with the activityModal
image

3.  
I have not been able to reproduce it anymore, can you confirm it?

**Comment by alfetopito** [comment]
> Now I've added a simple message, but maybe the most convenient is to be consistent with the activityModal and show the tx detail of the safe, after 150 multi-signature transactions 😆 I think this is very useful. WDYT @alfetopito, @ramirotw? image image
> 

Agreed!

**Comment by elena-zh** [comment]
Hey @henrypalacios , nice!

I have found several related issues.. And some issues for discussion. 
1. When do not sign a transaction immediately, 'pending' transaction appears in the activity modal. It appears even if you cancel signing (that is why #229 task exists on the board). In the current implementation it affects progress bar tooltip, due to it appears when at the moment 'pending' transaction appears in the activity modal. And here is an issue: Order confirmation modal is opened, and tooltip is displayed: 
Screenshot_1
Could we fix tooltip behavior? It can't be a blocker for merging this PR and can't be addressed in the next iteration. Let me know if I need to create an issue for this. 
2. The block does not look good to me inside the modal. Sees like it is too small, and there is a lot of room on the modal. Can we enhance it somehow? Maybe not display it as a block, remove boarders, etc. WDYT? cc @alfetopito , @alongoni 
maybe remove boarders
3. it is related to the #795 issue. It affects the current pr in terms of the progress bar: it appears when the order goes to 'Open' status. Could we please fix it somehow? 
4. The message about 'cancelling' order can't be applicable to a SC-wallet. Can we rewrite it somehow/not to show it when connected to a SC-wallet? @alfetopito , WDYT? 

Thanks!

**Comment by elena-zh** [comment]
Hey @henrypalacios , I can't reproduce case 1 today, so let not waist time on it. 
Case 4 LGTM!
Case 2 looks better, but maybe we could add a bit bigger margin between the boarder and the 1st message in a mobile view. WDYT?
not so close

Thanks!

**Review by elena-zh — APPROVED** [review]
I'm approving this PR in order not to block merging it to Develop. 
However, there are 2 related issues that would be great to address in the next iteraiton: 
 - No 'Price out of market' message when connected to a SC-wallet https://github.com/cowprotocol/cowswap/issues/865
 - SC wallets: order is on 'Open' status when it is expired  https://github.com/cowprotocol/cowswap/issues/795

File: src/custom/state/gnosisSafe/updater.tsx [hunk 4]
```diff
@@ -20,3 +27,81 @@ export default function Updater(): null {
...
+
+  const listenerEvent = useCallback(
+    async (txHash: string, _payment: BigNumber, event: TransactionResponse) => {
+      if (!chainId || !provider) return
+      const safeOrders = _getSafeTxs(pendingTxsRef.current)
+      let executionDate = new Date()
+      try {
+        executionDate = await getBlockTime(provider, event.blockNumber as number)
+      } catch (error) {
+        console.error(`Unable to obtain when mining block ${event.blockNumber}`, error)
```
Comment (alfetopito) [inline src/custom/state/gnosisSafe/updater.tsx:73]: Nitpick

```suggestion
        console.error(`Unable to fetch block ${event.blockNumber} time`, error)
```

**Review by alfetopito — APPROVED** [review]
Looks good, Henry!

I suggest you to merge it to a new feature branch based off of develop to aggregate all of the remaining points raised by Elena.

Could be `smart-contract-progress-bar` or something like that

**Pull request merged**